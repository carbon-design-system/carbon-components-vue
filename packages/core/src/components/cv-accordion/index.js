import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
