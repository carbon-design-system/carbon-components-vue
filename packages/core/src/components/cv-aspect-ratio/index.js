import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
