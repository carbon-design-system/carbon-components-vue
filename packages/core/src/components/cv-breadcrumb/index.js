import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
