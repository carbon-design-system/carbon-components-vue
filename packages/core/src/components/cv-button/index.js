import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
