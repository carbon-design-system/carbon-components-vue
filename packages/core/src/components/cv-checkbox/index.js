import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
