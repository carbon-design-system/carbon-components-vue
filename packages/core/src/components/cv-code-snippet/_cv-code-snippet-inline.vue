<template>
  <cv-feedback-button
    v-if="!hideCopyButton"
    class="cv-code-snippet-inline"
    :class="classes"
    :disabled="disabled"
    :feedback="copyFeedback"
    inline
    :aria-label="feedbackAriaLabel"
    @click="$emit('copy-code')"
    ><slot
  /></cv-feedback-button>
  <span v-else :class="classes"><slot /></span>
</template>

<script>
import CvFeedbackButton from '../cv-feedback-button/_cv-feedback-button';
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvCodeSnippetInline',
  mixins: [carbonPrefixMixin],
  components: {
    CvFeedbackButton,
  },
  props: {
    copyFeedback: String,
    disabled: Boolean,
    feedbackAriaLabel: String,
    hideCopyButton: Boolean,
  },
  computed: {
    classes() {
      return [`${this.carbonPrefix}--snippet`, `${this.carbonPrefix}--snippet--inline`];
    },
  },
};
</script>
