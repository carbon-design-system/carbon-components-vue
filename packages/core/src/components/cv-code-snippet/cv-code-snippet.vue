<template>
  <component
    class="cv-code-snippet"
    :is="theComponent"
    v-bind="$attrs"
    @copy-code="onCopyCode"
    :copy-feedback="copyFeedback"
    :feedback-aria-label="feedbackAriaLabel"
  >
    <code ref="code">
      <slot></slot>
    </code>
    <!-- textarea cannot be fully hidden for clipboard to work -->
    <textarea
      class="cv-code-snippet__clippy"
      style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"
      aria-hidden="true"
      ref="clippy"
    ></textarea>
  </component>
</template>

<script>
import CvCodeSnippetInline from './_cv-code-snippet-inline';
import CvCodeSnippetMultiline from './_cv-code-snippet-multiline';
import CvCodeSnippetOneline from './_cv-code-snippet-oneline';

export default {
  name: 'CvCodeSnippet',
  inheritAttrs: false,
  components: {
    CvCodeSnippetInline,
    CvCodeSnippetMultiline,
    CvCodeSnippetOneline,
  },
  props: {
    feedbackAriaLabel: { type: String, default: 'Copy code' },
    copyFeedback: { type: String, default: 'Copied!' },
    kind: {
      type: String,
      default: 'oneline',
      validator: value => ['inline', 'multiline', 'oneline'].includes(value),
    },
  },
  computed: {
    theComponent() {
      switch (this.kind) {
        case 'inline':
          return 'CvCodeSnippetInline';
        case 'multiline':
          return 'CvCodeSnippetMultiline';
        default:
          return 'CvCodeSnippetOneline';
      }
    },
  },
  methods: {
    onCopyCode() {
      // copy to clipboard
      this.$refs.clippy.value = this.$refs.code.innerText;
      this.$refs.clippy.select();
      document.execCommand('copy');
    },
  },
};
</script>
