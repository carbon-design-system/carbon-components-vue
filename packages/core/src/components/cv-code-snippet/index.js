import CvCodeSnippet from './cv-code-snippet';
import CvCodeSnippetSkeleton from './cv-code-snippet-skeleton';
export { CvCodeSnippet, CvCodeSnippetSkeleton };
