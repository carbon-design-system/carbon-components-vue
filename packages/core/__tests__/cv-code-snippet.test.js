import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger, mount } = awaitNextTick;
import { CvCodeSnippet, CvCodeSnippetSkeleton } from '@/components/cv-code-snippet';

describe('CvCodeSnippet', () => {
  const kinds = ['inline', 'multiline', 'oneline'];

  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvCodeSnippet, ['feedbackAriaLabel', 'copyFeedback', 'kind'], String);
  testComponent.propsHaveDefault(CvCodeSnippet, ['feedbackAriaLabel', 'copyFeedback', 'kind']);

  it('`kind` prop validator works as expected', async () => {
    const wrapper = await shallow(CvCodeSnippet);
    for (const kind of kinds) {
      expect(wrapper.vm.$options.props.kind.validator && wrapper.vm.$options.props.kind.validator(kind)).toBeTruthy();
    }

    // suppress the error message from the kind validator
    const consoleError = console.error;
    console.error = jest.fn();

    expect(wrapper.vm.$options.props.kind.validator && wrapper.vm.$options.props.kind.validator('TEST')).toBeFalsy();

    // restore
    console.error = consoleError;
  });

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('should render with default props correctly', async () => {
    const propsData = { id: 'test-1' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render with set `feedbackAriaLabel` and `copyFeedback` props correctly', async () => {
    const propsData = {
      id: 'test-1',
      feedbackAriaLabel: 'feedback aria label test',
      copyFeedback: 'copy feedback test',
    };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when it is `inline`', async () => {
    const propsData = { id: 'test-1', kind: 'inline' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when it is `multiline`', async () => {
    const propsData = { id: 'test-1', kind: 'multiline' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly with slot specified', async () => {
    const propsData = { id: 'test-1' };
    const wrapper = await shallow(CvCodeSnippet, {
      slots: {
        default: '<div>Cv code snippet slot</div>',
      },
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('should compute `theComponent` correctly when it is `oneline`', async () => {
    const propsData = { id: 'test-1', kind: 'oneline' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.vm.theComponent).toEqual('CvCodeSnippetOneline');
  });

  it('should compute `theComponent` correctly when it is `inline`', async () => {
    const propsData = { id: 'test-1', kind: 'inline' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.vm.theComponent).toEqual('CvCodeSnippetInline');
  });

  it('should compute `theComponent` correctly when it is `multiline`', async () => {
    const propsData = { id: 'test-1', kind: 'multiline' };
    const wrapper = await shallow(CvCodeSnippet, { propsData });

    expect(wrapper.vm.theComponent).toEqual('CvCodeSnippetMultiline');
  });

  it('should copy code to the clipboard on `copy-code` event', async () => {
    const execCommand = document.execCommand;
    document.execCommand = jest.fn();

    const propsData = { id: 'test-1', kind: 'multiline' };

    const wrapper = await mount(CvCodeSnippet, {
      slots: {
        default: '<div>Cv code snippet slot</div>',
      },
      propsData,
    });
    await trigger(wrapper.find('button'), 'click');
    expect(document.execCommand).toHaveBeenCalledWith('copy');

    // restore
    document.execCommand = execCommand;
  });
});

describe('CvCodeSnippetSkeleton', () => {
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvCodeSnippetSkeleton, ['kind'], String);
  testComponent.propsHaveDefault(CvCodeSnippetSkeleton, ['kind']);

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('should render with default props correctly', async () => {
    const propsData = { id: 'test-1' };
    const wrapper = await shallow(CvCodeSnippetSkeleton, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when it is `multiline`', async () => {
    const propsData = { id: 'test-1', kind: 'multiline' };
    const wrapper = await shallow(CvCodeSnippetSkeleton, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('should compute `multi` correctly when it is `oneline`', async () => {
    const propsData = { id: 'test-1', kind: 'oneline' };
    const wrapper = await shallow(CvCodeSnippetSkeleton, { propsData });

    expect(wrapper.vm.multi).toBeFalsy();
  });

  it('should compute `multi` correctly when it is `multiline`', async () => {
    const propsData = { id: 'test-1', kind: 'multiline' };
    const wrapper = await shallow(CvCodeSnippetSkeleton, { propsData });

    expect(wrapper.vm.multi).toBeTruthy();
  });
});
