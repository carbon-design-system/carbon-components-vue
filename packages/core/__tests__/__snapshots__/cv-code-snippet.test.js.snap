// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvCodeSnippet should render correctly when it is \`inline\` 1`] = `<cvcodesnippetinline-stub copyfeedback="Copied!" feedbackarialabel="Copy code" id="test-1" class="cv-code-snippet"><code></code> <textarea aria-hidden="true" class="cv-code-snippet__clippy" style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"></textarea></cvcodesnippetinline-stub>`;

exports[`CvCodeSnippet should render correctly when it is \`multiline\` 1`] = `<cvcodesnippetmultiline-stub copyfeedback="Copied!" feedbackarialabel="Copy code" lesstext="Show less" moretext="Show more" id="test-1" class="cv-code-snippet"><code></code> <textarea aria-hidden="true" class="cv-code-snippet__clippy" style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"></textarea></cvcodesnippetmultiline-stub>`;

exports[`CvCodeSnippet should render correctly with slot specified 1`] = `<cvcodesnippetoneline-stub copyfeedback="Copied!" feedbackarialabel="Copy code" id="test-1" class="cv-code-snippet"><code><div>Cv code snippet slot</div></code> <textarea aria-hidden="true" class="cv-code-snippet__clippy" style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"></textarea></cvcodesnippetoneline-stub>`;

exports[`CvCodeSnippet should render with default props correctly 1`] = `<cvcodesnippetoneline-stub copyfeedback="Copied!" feedbackarialabel="Copy code" id="test-1" class="cv-code-snippet"><code></code> <textarea aria-hidden="true" class="cv-code-snippet__clippy" style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"></textarea></cvcodesnippetoneline-stub>`;

exports[`CvCodeSnippet should render with set \`feedbackAriaLabel\` and \`copyFeedback\` props correctly 1`] = `<cvcodesnippetoneline-stub copyfeedback="copy feedback test" feedbackarialabel="feedback aria label test" id="test-1" class="cv-code-snippet"><code></code> <textarea aria-hidden="true" class="cv-code-snippet__clippy" style="position: absolute; left: -9999px; max-width: 0; opacity: 0; overflow: hidden;"></textarea></cvcodesnippetoneline-stub>`;

exports[`CvCodeSnippetSkeleton should render correctly when it is \`multiline\` 1`] = `
<div class="bx--snippet bx--skeleton bx--snippet--multi" id="test-1">
  <div class="bx--snippet-container"><span></span> <span></span> <span></span></div>
</div>
`;

exports[`CvCodeSnippetSkeleton should render with default props correctly 1`] = `
<div class="bx--snippet bx--skeleton bx--snippet--single" id="test-1">
  <div class="bx--snippet-container"><span></span>
    <!---->
    <!---->
  </div>
</div>
`;
